hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/token/delegation/ZKDelegationTokenSecretManager.java

```diff
@@ -363,105 +357,122 @@ public void startThreads() throws IOException {
-                case CHILD_UPDATED:
-                  processTokenAddOrUpdate(event.getData().getData());
-                  break;
-                case CHILD_REMOVED:
-                  processTokenRemoved(event.getData());
-                  break;
-                default:
-                  break;
-              }
-            }
-          }, listenerThreadPool);
+          CuratorCacheListener tokenCacheListener = CuratorCacheListener.builder()
+              .forCreates(childData -> {
+                try {
+                  processTokenAddOrUpdate(childData.getData());
+                } catch (IOException e) {
+                  LOG.error("Error while processing Curator tokenCacheListener "
+                      + "NODE_CREATED event");
+                  throw new UncheckedIOException(e);
+                }
+              })
+              .forChanges((oldNode, node) -> {
+                try {
+                  processTokenAddOrUpdate(node.getData());
+                } catch (IOException e) {
+                  LOG.error("Error while processing Curator tokenCacheListener "
+                      + "NODE_CHANGED event");
+                  throw new UncheckedIOException(e);
+                }
+              })
+              .forDeletes(childData -> {
+                try {
+                  processTokenRemoved(childData);
+                } catch (IOException e) {
+                  LOG.error("Error while processing Curator tokenCacheListener "
+                      + "NODE_DELETED event");
+                  throw new UncheckedIOException(e);
+                }
+              })
+              .build();
+          tokenCache.listenable().addListener(tokenCacheListener);
+          tokenCache.start();
           loadFromZKCache(true);
         }
       } catch (Exception e) {
-        throw new IOException("Could not start PathChildrenCache for tokens", e);
+        throw new IOException(
+            "Could not start Curator tokenCacheListener for tokens", e);
       }
     }
     super.startThreads();
   }
 
   /**
-   * Load the PathChildrenCache into the in-memory map. Possible caches to be
+   * Load the CuratorCache into the in-memory map. Possible caches to be
    * loaded are keyCache and tokenCache.
    *
    * @param isTokenCache true if loading tokenCache, false if loading keyCache.
    */
   private void loadFromZKCache(final boolean isTokenCache) {
     final String cacheName = isTokenCache ? "token" : "key";
     LOG.info("Starting to load {} cache.", cacheName);
-    final List<ChildData> children;
+    final Stream<ChildData> children;
     if (isTokenCache) {
-      children = tokenCache.getCurrentData();
+      children = tokenCache.stream();
     } else {
-      children = keyCache.getCurrentData();
+      children = keyCache.stream();
     }
 
-    int count = 0;
-    for (ChildData child : children) {
+    final AtomicInteger count = new AtomicInteger(0);
+    children.forEach(childData -> {
```

Comment:
I read the Curator's source code, and found that `children` is not parallel. Should we make it parallel by `children.parallel().forEach()` or not?

Reply:
I think it's fine even with serial (as we were doing before this change in for loop), but in case you are wondering why we are using AtomicInteger here, it's because lambda allows only final or eventually final variables inside the block.

Reply:
@aajisaka Does this sound good to you? If you feel strong about benefits of using Stream parallel API here, I can give it a shot. I just wanted to keep the behaviour same since we were already looping on children serially before.

Reply:
It sounds good for me. Thank you!